src/uct/cuda/cuda_ipc/cuda_ipc_cache.c

```diff
@@ -391,19 +388,17 @@ void uct_cuda_ipc_destroy_cache(uct_cuda_ipc_cache_t *cache)
 }
 
 UCS_STATIC_INIT {
-    ucs_recursive_spinlock_init(&uct_cuda_ipc_map_lock, 0);
-    ucs_recursive_spinlock_init(&remote_cache.lock, 0);
-    kh_init_inplace(cuda_ipc_rem_cache, &remote_cache.hash);
+    ucs_recursive_spinlock_init(&uct_cuda_ipc_remote_cache.lock, 0);
```

Comment:
`uct_cuda_ipc_map_lock` is to have mutual exclusion for threads trying open an IPC memory handle whereas `uct_cuda_ipc_remote_cache.lock` is used to ensure that no two threads try to access the hashtable of cuda_ipc cache structures created per remote process per local cuda device.

Reply:
why do we need uct_cuda_ipc_map_lock ? isn't cuda API already thread safe?

Reply:
@Akshay-Venkatesh we dont need uct_cuda_ipc_map_lock() right? driver seems already takes a cuda context lock

Reply:
@bureddy @yosefe It's safe for multiple threads to call cuIpcOpenMemHandle. We still want to avoid the case where two threads end up checking for reachability and end up causing "memory handle already opened error". For this we need mutual exclusion on the [segment of code](https://github.com/Akshay-Venkatesh/ucx/blob/topic/cuda-ipc-iface-cache/src/uct/cuda/cuda_ipc/cuda_ipc_md.c#L149) that 1. opens the remote memory handle 2. updates peer_accessibility table 3. closes the memory handle. For this reason when need to use a lock around the above segment but it does seem like the same lock is not needed [here](https://github.com/Akshay-Venkatesh/ucx/blob/topic/cuda-ipc-iface-cache/src/uct/cuda/cuda_ipc/cuda_ipc_cache.c#L110) because by that time the thread is guaranteed to have checked for reachability. I've removed the lock use in `uct_cuda_ipc_open_memhandle` for this reason.